libedgetpu: package and provide nixos module

**Project description**

It would be nice to be able to use the Google Coral Edge TPU Usb device with NixOS.

I think it's mostly a matter of building libedgetpu (and maybe tensorflow?) and then bundling the udev rules with it?

**Metadata**

* homepage URL:  https://github.com/google-coral/libedgetpu
* source URL: https://github.com/google-coral/libedgetpu
* license: apache 2
* platforms: linux, darwin

Is there a configuration example around on how to get this all working?

My goal is have Frigate running with a Coral mini-PCIe card, currently using the nixpkgs-23.11 versions of *gasket* and *frigate* and a custom

<details>
<summary>libedgetpu.nix</summary>

```nix
{ stdenv
, lib
, fetchFromGitHub
, python3
, libusb1
, abseil-cpp
, flatbuffers
, xxd
}:

let
  flatbuffers_1_12 = flatbuffers.overrideAttrs (oldAttrs: rec {
    version = "1.12.1";
    NIX_CFLAGS_COMPILE = "-Wno-error=class-memaccess -Wno-error=maybe-uninitialized";
    cmakeFlags = (oldAttrs.cmakeFlags or []) ++ ["-DFLATBUFFERS_BUILD_SHAREDLIB=ON"];
    NIX_CXXSTDLIB_COMPILE = "-std=c++17";
    configureFlags = (oldAttrs.configureFlags or []) ++ ["--enable-shared"];
    src = fetchFromGitHub {
      owner = "google";
      repo = "flatbuffers";
      rev = "v${version}";
      sha256 = "sha256-5sHddlqWx/5d5/FOMK7qRlR5xbUR47rfejuXI1jymWM=";
    };
  });

in stdenv.mkDerivation rec {
  # pname = "libedgetpu";
  # version = "ddfa7bde33c23afd8c2892182faa3e5b4e6ad94e";

  # src = fetchFromGitHub {
  #   owner = "google-coral";
  #   repo = pname;
  #   rev = version;
  #   sha256 = "sha256-NidGjBPOLu5py7bakqvNQLDi72b5ig9QF9C1UuQldn0=";
  # };

  pname = "libedgetpu";
  version = "90b03d96ed83412178ed6e6cfddbd40bb3f84925";

  src = fetchFromGitHub {
    owner = "feranick";
    repo = pname;
    rev = version;
    sha256 = "sha256-/Eneik+v+juGsg/us+0YBQxkKeJUpGnFqrRPu5nKYWk=";
  };

  # patches = [ ./libedgetpu-stddef.patch ];

  makeFlags = ["-f" "makefile_build/Makefile" "libedgetpu" ];

  buildInputs = [
    libusb1
    abseil-cpp
    flatbuffers_1_12
  ];

  nativeBuildInputs = [
    xxd
  ];

  NIX_CXXSTDLIB_COMPILE = "-std=c++17";

  TFROOT = "${fetchFromGitHub {
    owner = "tensorflow";
    repo = "tensorflow";
    rev = "v2.8.4"; # latest rev providing tensorflow/lite/c/common.c
    sha256 = "sha256-MFqsVdSqbNDNZSQtCQ4/4DRpJPG35I0La4MLtRp37Rk=";
    # rev = "v2.13.1"; # latest rev providing tensorflow/lite/c/common.c
    # sha256 = "sha256-fCwf7I76gdyeOPVnPPqEw4cI7RrcrshTSHjdfevUriY=";
  }}";

  # TFROOT = "${python3.pkgs.tensorflow}";

  enableParallelBuilding = false;

  installPhase = ''
    mkdir -p $out/lib
    cp out/direct/k8/libedgetpu.so.1.0 $out/lib
    ln -s $out/lib/libedgetpu.so.1.0 $out/lib/libedgetpu.so.1
    mkdir -p $out/lib/udev/rules.d
    cp debian/edgetpu-accelerator.rules $out/lib/udev/rules.d/99-edgetpu-accelerator.rules
  '';
}
```
</details>

Injecting it into Frigate like

```nix
      systemd.services.frigate.environment.LD_LIBRARY_PATH = "${libedgetpu}/lib";
```

Best I could do so far was a successful build when using TensorFlow 2.8.4, newer versions result in

```
make: *** No rule to make target '/nix/store/ik160yl79rjszcdgj301apza3a5vhd0z-source/tensorflow/lite/c/common.c', needed by '/build/source/makefile_build/../out//nix/store/ik160yl79rjszcdgj301apza3a5vhd0z-source/tensorflow/lite/c/common.o'.  Stop.
```

which should be [fixed in 2.13.1](https://github.com/tensorflow/tensorflow/pull/54566), but no difference. nixpkgs-23.11 includes TF 2.13.0, that is what I was aiming at.

With 2.8.4 Frigate produces a

<details>
<summary>segfault</summary>

```
frigate.detectors.plugins.edgetpu_tfl INFO    : Attempting to load TPU as pci
frigate.app                    INFO    : Output process started: 69432
frigate.app                    INFO    : Camera processor started for test2: 69437
frigate.app                    INFO    : Capture process started for test2: 69441
Fatal Python error: Segmentation fault
Thread 0x00007f0ddca4e6c0 (most recent call first):
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/threading.py", line 327 in wait
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/queues.py", line 231 in _feed
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/threading.py", line 982 in run
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/threading.py", line 1045 in _bootstrap_inner
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/threading.py", line 1002 in _bootstrap
Current thread 0x00007f0e31654740 (most recent call first):
  File "/nix/store/grc5rgqywnlgjm40w0kb8kz0n95vkn4b-python3.11-tensorflow-2.13.0/lib/python3.11/site-packages/tensorflow/lite/python/interpreter.py", line 513 in __init__
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/detectors/plugins/edgetpu_tfl.py", line 39 in __init__
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/detectors/__init__.py", line 24 in create_detector
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/object_detection.py", line 52 in __init__
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/object_detection.py", line 98 in run_detector
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/process.py", line 108 in run
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/process.py", line 314 in
frigate.detectors.plugins.edgetpu_tfl INFO    : TPU found
_bootstrap
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/popen_fork.py", line 71 in _launch
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/popen_fork.py", line 19 in __init__
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/context.py", line 281 in _Popen
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/context.py", line 224 in _Popen
  File "/nix/store/rac8pxbi1vapwrlqzbrkycbyg521djzw-python3-3.11.6/lib/python3.11/multiprocessing/process.py", line 121 in start
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/object_detection.py", line 179 in start_or_restart
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/object_detection.py", line 147 in __init__
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/app.py", line 214 in start_detectors
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/app.py", line 379 in start
  File "/nix/store/l0gfxyfahpx92v00ylwavhpgzd05s296-frigate-0.12.1/lib/python3.11/site-packages/frigate/__main__.py", line 16 in <module>
  File "<frozen runpy>", line 88 in _run_code
  File "<frozen runpy>", line 198 in _run_module_as_main
```
</details>

Also I was surprised that [the frigate package](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/pkgs/applications/video/frigate/default.nix#L129) depends on *tensorflow* instead of *tensorflow-lite*.

Thank you for the effort!

--- a/maintainers/maintainer-list.nix
+++ b/maintainers/maintainer-list.nix
@@ -7188,6 +7188,11 @@
     githubId = 13279982;
     name = "freezeboy";
   };
+  frenetic00 = {
+    github = "frenetic00";
+    githubId = 50942055;
+    name = "frenetic00";
+  };
   Fresheyeball = {
     email = "fresheyeball@gmail.com";
     github = "Fresheyeball";

--- a/pkgs/by-name/li/libedgetpu/package.nix
+++ b/pkgs/by-name/li/libedgetpu/package.nix
@@ -1,0 +1,87 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  libusb1,
+  abseil-cpp,
+  flatbuffers,
+  fetchpatch,
+  xxd,
+}:
+let
+  # Tensorflow 2.16.1 requires Flatbuffers 23.5.26
+  # Compile as a shared library
+  flatbuffers_23_5_26 = flatbuffers.overrideAttrs (oldAttrs: rec {
+    version = "23.5.26";
+    cmakeFlags = (oldAttrs.cmakeFlags or [ ]) ++ [ "-DFLATBUFFERS_BUILD_SHAREDLIB=ON" ];
+    NIX_CXXSTDLIB_COMPILE = "-std=c++17";
+    configureFlags = (oldAttrs.configureFlags or [ ]) ++ [ "--enable-shared" ];
+    src = fetchFromGitHub {
+      owner = "google";
+      repo = "flatbuffers";
+      rev = "v${version}";
+      hash = "sha256-e+dNPNbCHYDXUS/W+hMqf/37fhVgEGzId6rhP3cToTE=";
+    };
+  });
+in
+stdenv.mkDerivation {
+  pname = "libedgetpu";
+  version = "0-unstable-2024-03-14";
+
+  src = fetchFromGitHub {
+    owner = "google-coral";
+    repo = "libedgetpu";
+    rev = "e35aed18fea2e2d25d98352e5a5bd357c170bd4d";
+    hash = "sha256-SabiFG/EgspiCFpg8XQs6RjFhrPPUfhILPmYQQA1E2w=";
+  };
+
+  patches = [
+    (fetchpatch {
+      name = "fix-makefile-to-compile-with-latest-tensorflow.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/google-coral/libedgetpu/pull/66.patch";
+      hash = "sha256-mMODpQmikfXtsQvtgh26cy97EiykYNLngSjidOBt/3I=";
+    })
+  ];
+
+  makeFlags = [
+    "-f"
+    "makefile_build/Makefile"
+    "libedgetpu"
+  ];
+
+  buildInputs = [
+    abseil-cpp
+    libusb1
+    flatbuffers_23_5_26
+  ];
+
+  nativeBuildInputs = [ xxd ];
+
+  NIX_CXXSTDLIB_COMPILE = "-std=c++17";
+
+  TFROOT = fetchFromGitHub {
+    owner = "tensorflow";
+    repo = "tensorflow";
+    rev = "v2.16.1";
+    hash = "sha256-UPvK5Kc/FNVJq3FchN5IIBBObvcHtAPVv0ARzWzA35M=";
+  };
+
+  enableParallelBuilding = true;
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm555 out/direct/k8/libedgetpu.so.1.0 -t $out/lib
+    ln -s $out/lib/libedgetpu.so.1.0 $out/lib/libedgetpu.so.1
+    install -Dm444 debian/edgetpu-accelerator.rules $out/lib/udev/rules.d/99-edgetpu-accelerator.rules
+    install -Dm444 tflite/public/*.h -t $out/include
+    runHook postInstall
+  '';
+
+  meta = {
+    homepage = "https://github.com/google-coral/libedgetpu";
+    description = "Userspace level runtime driver for Coral devices";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ frenetic00 ];
+    platforms = lib.platforms.linux;
+  };
+}

